### PR TITLE
[Antha-2258]/serialization between antha-lang and e.g. PipetMaxDriver

### PR DESCRIPTION
**Pull request by haydnKing**
This PR is a significant rewrite of the serialisation that occurs in order to communicate with instruction plugins.

The goal of the change is to make serialisation more robust to changes in structs which describe robots or physical objects within robots. This is achieved by removing custom serialization code where possible, and where not making use of custom JSON serialization such that it sits within the same package as the struct itself.

The library `drivers/liquidhandling` wraps the creation and invocation of the GRPC server, and handles conversion into protobuf defined messages for non-complex types.

More complex objects (e.g. wtype.LHPlate, liquidhandling.LHProperties), are sent using JSON serialisation - for many objects (plates, tipboxes) this already existed and needed maintaining, so adds no extra burden. Serialisation was added for LHProperties, being careful to make sure that pointers refer to the correct underlying object in the deserialised object.

### CODE REVIEW

File: driver/liquidhandling/server_test.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+				drv.AddPlateTo("position_1", makePlateForTest(), "firstPlate")
+				drv.AddPlateTo("position_2", makePlateForTest(), "secondPlate")
+				drv.Transfer([]string{"the crown jewels"}, []string{"london"}, []string{"tower of"}, []string{"me"}, []string{"head"}, []float64{100.0})
+				drv.Message(100, "all your joules", "are belong to me", false)
+				drv.Finalize()
+			},
+			Expected: []string{
+				"Initialize()",
+				"GetCapabilities()",
+				"AddPlateTo(\"position_1\", *wtype.Plate, \"firstPlate\")",
```
Comment (m-i-sadowski) [inline driver/liquidhandling/server_test.go]: Should be testing the plate is at least in some way correct here?

File: driver/liquidhandling/server_test.go [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+				drv.Dispense([]float64{100.0}, []bool{false}, 0, 1, []string{"plate"}, []string{"wine"}, []bool{false})
+				drv.Move([]string{"position_1"}, []string{"A1"}, []int{1}, []float64{0.0}, []float64{0.0}, []float64{5.0}, []string{"tipbox"}, 0)
+				drv.UnloadTips([]int{0}, 0, 1, []string{"tipbox"}, []string{"position_2"}, []string{"A1"})
+				drv.Message(100, "from water", "into wine", false)
+				drv.Finalize()
+			},
+			Expected: []string{
+				"Initialize()",
+				"GetCapabilities()",
+				"AddPlateTo(\"position_1\", *wtype.LHTipwaste, \"tipwaste\")",
```
Comment (m-i-sadowski) [inline driver/liquidhandling/server_test.go]: Painful as it is I do think there's real value in checking if the plates etc are coming out the same as they went in here...

File: microArch/driver/liquidhandling/lhproperties.go [hunk 3]
```diff
@@ -42,28 +43,25 @@ import (
...
-	Mnfr          string
-	LHType        string
-	TipType       string
-	//Heads lists every head (whether loaded or not) that is available for the machine
-	Heads []*wtype.LHHead
-	//Adaptors lists every adaptor (whether loaded or not) that is available for the machine
-	Adaptors []*wtype.LHAdaptor
-	//HeadAssemblies describes how each loaded head and adaptor is loaded into the machine
-	HeadAssemblies       []*wtype.LHHeadAssembly
+	ID                   string
```
Comment (m-i-sadowski) [inline microArch/driver/liquidhandling/lhproperties.go]: Push back if you wish but I can't help but feeling there's an opportunity to remove some unused fields here (e.g. Devices)

File: microArch/driver/liquidhandling/lhproperties.go [hunk 4]
```diff
@@ -42,28 +43,25 @@ import (
...
+	PlateLookup          map[string]interface{}       // deck object (plate, tipbox, etc) by object ID
+	PosLookup            map[string]string            // object ID by position name
+	PlateIDLookup        map[string]string            // position name by object ID
+	Plates               map[string]*wtype.Plate      // plates by position name
+	Tipboxes             map[string]*wtype.LHTipbox   // tipboxes by position name
+	Tipwastes            map[string]*wtype.LHTipwaste // tipwastes by position name
+	Wastes               map[string]*wtype.Plate      // waste plates by position name
+	Washes               map[string]*wtype.Plate      // wash plates by position name
+	Devices              map[string]string
+	Model                string
```
Comment (m-i-sadowski) [inline microArch/driver/liquidhandling/lhproperties.go]: Also a comment on what the fields mean would help

File: microArch/driver/liquidhandling/lhproperties_test.go [hunk 5]
```diff
@@ -186,3 +188,56 @@ func AssertLHPropertiesEqual(t *testing.T, e, g *LHProperties, msg string) {
...
+
+	var after LHProperties
+	if data, err := json.Marshal(before); err != nil {
+		t.Error(err)
+	} else if err := json.Unmarshal(data, &after); err != nil {
+		t.Error(err)
+	}
+
+	if !reflect.DeepEqual(before, &after) {
+		t.Errorf("serialization changed LHProperties PlateLookup\ne: %+v\ng: %+v", before, &after)
```
Comment (m-i-sadowski) [inline microArch/driver/liquidhandling/lhproperties_test.go]: Again more explicit "before / after" here would be a bit nicer  ("e/g issue")

File: microArch/driver/liquidhandling/lhproperties_test.go [hunk 6]
```diff
@@ -186,3 +188,56 @@ func AssertLHPropertiesEqual(t *testing.T, e, g *LHProperties, msg string) {
...
+		adaptors[a] = true
+	}
+	for _, head := range after.Heads {
+		if head.Adaptor != nil && !adaptors[head.Adaptor] {
+			t.Error("Head.Adaptor doesn't point to anything in LHProperties.Adaptors")
+		}
+	}
+
+	if e, g := before.GetLoadedHeads(), after.GetLoadedHeads(); len(e) != len(g) {
+		t.Errorf("number of loaded heads doesn't match: e: %d, g: %d", len(e), len(g))
```
Comment (m-i-sadowski) [inline microArch/driver/liquidhandling/lhproperties_test.go]: e / g issue again

**Comment by m-i-sadowski** [comment]
I've tried this with two protocols on the pipetmax driver and there is a bug which seems to be confusing tipwastes with tip boxes

screen shot 2018-11-13 at 11 45 57
screen shot 2018-11-13 at 11 46 10

File: antha/anthalib/wtype/serialize.go [hunk 7]
```diff
@@ -25,8 +25,52 @@ package wtype
 import (
 	"encoding/json"
 	"github.com/antha-lang/antha/antha/anthalib/wunit"
+	"github.com/pkg/errors"
 )
 
+// MarshalDeckObject marshals any object which can be placed on a deck into valid JSON
```
[the remark this answers is not in the document]
Comment (haydnKing) [reply antha/anthalib/wtype/serialize.go]: I'm not sure I understand this one - what do you mean?

File: microArch/driver/liquidhandling/lhproperties.go (hunk 3 above)
Comment (haydnKing) [reply microArch/driver/liquidhandling/lhproperties.go]: I've removed unused fields and added types and comments to others where appropriate

**Comment by haydnKing** [comment]
Ok, I believe I've addressed everything and have run a bundle through trilution to verify. @m-i-sadowski you were spot on with the Load/Unload - I was missing an `un`

**Comment by m-i-sadowski** [comment]
I can confirm I've tested this again with some more complex bundles, all seems good to me.

**Pull request merged**